### DISCUSSION TITLE
Second Component Not Re-rendering After Dispatch

### DISCUSSION THREAD

**Discussion by cgunnels** [livewire/livewire#8038] [1. Help] [answered]
Running Laravel v10.45.0 and Livewire v3.4.6

I've got a search bar in a component called SecondHeader in which it searches products in the db and then sends those products to the dispatch. 
The second-header blade file:
```
<input type="search" id="default-search" wire:model.live.debounce.500ms="search" placeholder="Search products"/>
```

SecondHeader component
```
#[Session]
public string $search = '';

public function render()
{
   // search method is using laravel's scout library
    $products = Product::search($this->search)->paginate(5);
    $this->dispatch('refresh-products', $products);
    return view('livewire.second-header');
}
```

On my ProductList component i'm listening for the event
```
#[On('refresh-products')]
public function render($products = null): View
{
    if ($products === null) {
        $products = Product::all()->paginate(5);
    }
    return view('livewire.products.list', ['products' => $products]);
}
 ```
 If I `dd()` in the ProductList render method it shows the searched products but for some reason it won't update the blade file.
 
 Here's my blade file (I removed most table cells for brevity):
 ```
 <div class="flex-col space-y-4">
    <div>
        <x-table>
            <x-slot name="head">
                <x-table.heading sortable>Product Name</x-table.heading>
            </x-slot>
            <x-slot name="body">
                @forelse($products as $product)
                    <x-table.row wire:key="product-{{ $product->id }}">
                        <x-table.cell>{{ $product->title }}</x-table.cell>
                        <x-table.cell>
                            <a href="#" class="text-indigo-600 hover:text-indigo-900">Edit</a>
                        </x-table.cell>
                    </x-table.row>
                @empty
                    No products found
                @endforelse
            </x-slot>
        </x-table>
    </div>
    <div>
        {{ $products->links() }}
    </div>
</div>
 ```

**Comment by cgunnels** [comment]
Only took me 48 hrs to figure this out... although i can't get the pagination to work correctly i can atleast search now

I removed `$products = Product::search($this->search)->paginate(5);` from the second header and just sent the search string through because I found out that you can't pass objects from component to component. This makes sense but for some reason I was thinking it was php passing the object to another php class, but that isn't the case, it's obvious now.

I had to add a mount method and a __construct method for this to work. Here's my ProductList class:
```
...
use WithPagination, WithoutUrlPagination;

private LengthAwarePaginator $products;

public function __construct()
{
    $this->products = Product::where(...)->paginate(5);
}

public function mount()
{
    $this->products = Product::where(...)->paginate(5);
}

#[On('refresh-products')]
public function render($search = ''): View
{
    if (!empty($search)) {
        $this->products = Product::search($search)->paginate(5);
    }
    return view('livewire.products.list', ['products' => $this->products]);
}
```

**Reply by joshhanley** [reply, marked as answer]
@cgunnels you shouldn't be using a constructor with Livewire. The problem is you are trying to assign your paginated collection to a public property, but Livewire generally doesn't support that. Instead what you should do is use computed properties.

You are also dispatching Eloquent models around which is bad for performance and potentially a security concern.

I personally wouldn't have these in separate components as it makes things a lot more difficult. But assuming you have done it due to the layout (I would still make one component that wraps everything instead), I would do it like this:

For the search input, I would make it an Alpine component not a Livewire component:

```blade
<div x-data>
    <input type="text" x-on:input.debounce.500ms="$dispatch('search', {'value': $event.target.value})" />
</div>
```

Then in your products Livewire component

```php
// Make sure to import computed and On
use Livewire\Attributes\Computed;
use Livewire\Attributes\On; 

// Class
use WithPagination, WithoutUrlPagination;

public $search;

#[On('search')]
public function setSearch($value)
{
    $this->search = $value;
}

#[Computed]
public function products()
{
    if ($search) {
        return Product::search($search)->paginate(5);
    }
    
    return Product::paginate(5);
}

public function render(): View
{
    return view('livewire.products.list');
}
```

Then in your view, to access the products computed property, instead of `$products` you need to call `$this->products`

Like

```blade
...
    @foreach($this->products as $product)
...
```

Hope this helps!

**Reply by cgunnels** [reply]
Awesome thanks @joshhanley

**Reply by joshhanley** [reply]
@cgunnels no worries! 🙂
